prow/cmd/deck/job_history.go

```diff
@@ -0,0 +1,389 @@
+	ResultsShown int
+	ResultsTotal int
+	Builds       []buildData
+}
+
+func readObject(obj *storage.ObjectHandle) ([]byte, error) {
+	rc, err := obj.NewReader(context.Background())
+	if err != nil {
+		return []byte{}, fmt.Errorf("failed to get reader for GCS object: %v", err)
+	}
+	return ioutil.ReadAll(rc)
+}
+
+func readLatestBuild(bkt *storage.BucketHandle, root string) (int64, error) {
+	path := path.Join(root, latestBuildFile)
+	data, err := readObject(bkt.Object(path))
+	if err != nil {
+		return -1, fmt.Errorf("failed to read %s: %v", path, err)
+	}
+	n, err := strconv.ParseInt(string(data), 10, 64)
+	if err != nil {
+		return -1, fmt.Errorf("failed to parse %s: %v", path, err)
+	}
+	return n, nil
+}
+
+// resolve sym links into the actual log directory for a particular test run
+func resolveSymLink(bkt *storage.BucketHandle, symLink string) (string, error) {
+	data, err := readObject(bkt.Object(symLink))
+	if err != nil {
+		return "", fmt.Errorf("failed to read %s: %v", symLink, err)
+	}
+	// strip gs://<bucket-name> from global address `u`
+	u := string(data)
+	return prefixRe.ReplaceAllString(u, ""), nil
+}
+
+func spyglassLink(bkt *storage.BucketHandle, root, id string) (string, error) {
+	bAttrs, err := bkt.Attrs(context.Background())
+	if err != nil {
+		return "", fmt.Errorf("failed to get bucket name: %v", err)
+	}
+	bktName := bAttrs.Name
+	p, err := getPath(bkt, root, id, "")
+	if err != nil {
+		return "", fmt.Errorf("failed to get path: %v", err)
+	}
+	return path.Join(spyglassPrefix, bktName, p), nil
+}
+
+func getPath(bkt *storage.BucketHandle, root, id, fname string) (string, error) {
+	if strings.HasPrefix(root, logsPrefix) {
+		return path.Join(root, id, fname), nil
+	}
+	symLink := path.Join(root, id+".txt")
+	dir, err := resolveSymLink(bkt, symLink)
+	if err != nil {
+		return "", fmt.Errorf("failed to resolve sym link: %v", err)
+	}
+	return path.Join(dir, fname), nil
+}
+
+// reads specified JSON file in to `data`
+func readJSON(bkt *storage.BucketHandle, root, id, fname string, data interface{}) error {
+	p, err := getPath(bkt, root, id, fname)
+	if err != nil {
+		return fmt.Errorf("failed to get path: %v", err)
+	}
+	rawData, err := readObject(bkt.Object(p))
+	if err != nil {
+		return fmt.Errorf("failed to read %s for build %s: %v", fname, id, err)
+	}
+	err = json.Unmarshal(rawData, &data)
+	if err != nil {
+		return fmt.Errorf("failed to parse %s for build %s: %v", fname, id, err)
+	}
+	return nil
+}
+
+// Lists the GCS "directory names" immediately under prefix.
+func listSubDirs(bkt *storage.BucketHandle, prefix string) ([]string, error) {
+	if !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+	dirs := []string{}
+	it := bkt.Objects(context.Background(), &storage.Query{
+		Prefix:    prefix,
+		Delimiter: "/",
+	})
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return dirs, err
+		}
+		if attrs.Prefix != "" {
+			dirs = append(dirs, attrs.Prefix)
```

Comment:
Isn't `attrs.Prefix` the directory path not the directory name like the comment suggests? e.g. `foo/bar/baz` instead of `baz`.

Reply:
You're right. Updated